src/events/utilities/grade-calculator/grade-calculator.ts

```diff
@@ -1,41 +1,56 @@
-import { LowPassFilter } from './low-pass-filter';
+import { isNumber } from '../helpers';
+
+const KalmanFilter = require('kalmanjs');
+
+export const CLAMP = 40
 
 export class GradeCalculator {
 
-  public static computeGrade(previousDistance: number, currentDistance: number, previousAltitude: number, currentAltitude: number): number {
+  public static computeGrade(previousDistance: number | null, currentDistance: number | null, previousAltitude: number | null, currentAltitude: number | null): number {
+    previousDistance = previousDistance !== null ? previousDistance : 0;
+    currentDistance = currentDistance !== null ? currentDistance : 0;
+    previousAltitude = previousAltitude !== null ? previousAltitude : 0;
+    currentAltitude = currentAltitude !== null ? currentAltitude : 0;
+
     const distanceDelta = currentDistance - previousDistance;
     const altitudeDelta = currentAltitude - previousAltitude;
     if (distanceDelta === 0) {
       return 0;
     }
     let percentage: number = altitudeDelta / distanceDelta * 100;
-    percentage = Math.min(Math.max(percentage, -45), 45); // Clamp between -45% & 45%
+    percentage = Math.min(Math.max(percentage, -CLAMP), CLAMP); // Clamp between -CLAMP% & CLAMP%
     return Math.round(percentage * 10) / 10;
   }
 
-  public static computeGradeStream(distanceStream: number[], altitudeStream: number[]): number[] {
 
-    let gradeStream = [];
-
-    for (let i = 0; i < distanceStream.length; i++) {
+  // Find the next i for distance = prev distance + 5m
+  // @todo perhaps this threshold should be per avg speed / activity type
+  // Or just perhaps the speed?
+  // Perhaps round
+  public static computeGradeStream(
+    distanceStream: (number | null)[],
+    altitudeStream: (number | null)[],
+    filterAltitude = true,
+    filterGrade = true,
+    basedOnAltitude = true,
+    lookAhead = true
+  ): (number | null)[] {
+    // First filter the altitude to remove any noise and predict
+    if (filterAltitude) {
+      const kf = new KalmanFilter();
+      altitudeStream = altitudeStream.map(v => v === null ? null : kf.filter(v));
```

Comment:
Could this make sense to apply the filter directly on the altitude stream? I mean not just here for grade calculation. So when we call for instance `sportsLibActivity.getSquashedStreamData(DataAltitude.type);` we get the altitude array filtered also.

Reply:
Could be but then a user might loose the ability to debug his altitude problems (eg barometer broken etc). 
We could though just generate a new data type `DataFilteredAltitude` for a stream and use that if you would like to provide filtered data I assume.

Reply:
You right ;). The app using lib can also do the work.